File: 2025/SoyoungJung/week12/2025-08-24-Pacific-Atlantic-Water-Flow.ts
function pacificAtlantic(heights: number[][]): number[][] {
  const m = heights.length;
  const n = heights[0].length;

  const pacific = new Set<string>();
  const atlantic = new Set<string>();

  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];

  function dfs(r: number, c: number, visited: Set<string>, prevHeight: number) {
    if (r < 0 || c < 0 || r >= m || c >= n) return;
    const key = `${r},${c}`;
    if (visited.has(key)) return;
    if (heights[r][c] < prevHeight) return;

    visited.add(key);

    for (const [dr, dc] of directions) {
      dfs(r + dr, c + dc, visited, heights[r][c]);
    }
  }

  for (let c = 0; c < n; c++) {
    dfs(0, c, pacific, heights[0][c]);
    dfs(m - 1, c, atlantic, heights[m - 1][c]);
  }
  for (let r = 0; r < m; r++) {
    dfs(r, 0, pacific, heights[r][0]);
    dfs(r, n - 1, atlantic, heights[r][n - 1]);
  }

  const result: number[][] = [];
  for (let r = 0; r < m; r++) {
    for (let c = 0; c < n; c++) {
      const key = `${r},${c}`;
      if (pacific.has(key) && atlantic.has(key)) {
        result.push([r, c]);
      }
    }
  }

  return result;
}

File: 2025/Jaehyuk/2025-08-24-Pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:

        row, col = len(heights), len(heights[0])
        pac, atl = set(), set()

        def dfs(r, c, visit, prevHeight):
            if ((r, c) in visit or r < 0 or c < 0 or r==row or c==col or heights[r][c] < prevHeight):
                return
            visit.add((r,c))
            dfs(r+1, c, visit, heights[r][c])
            dfs(r-1, c, visit, heights[r][c])
            dfs(r, c+1, visit, heights[r][c])
            dfs(r, c-1, visit, heights[r][c])

        for c in range(col):
            dfs(0, c, pac, heights[0][c])
            dfs(row-1, c, atl, heights[row-1][c])

        for r in range(row):
            dfs(r, 0, pac, heights[r][0])
            dfs(r, col-1, atl, heights[r][col -1])
        res = []
        for r in range(row):
            for c in range(col):
                if (r,c) in pac and (r,c) in atl:
                    res.append([r,c])
        return res



                
                #check if each cell can reach both atlantic and pacific ocean
                # to check, check if left cell is less or equal to current cell and same to right cell
                # OR if atlantic ocean is on [1,0] to it or [0, -1]
                # ALSo check if pacific ocean is [-1,0] or [0,1] to it
